doc/designs/fast-param-find.md

```diff
@@ -0,0 +1,324 @@
+        switch (kdf_scrypt_ctx_lookup(p))
+        default:
+            break;
+
+        case kdf_scrypt_ctx_param_OSSL_KDF_PARAM_PASSWORD:
+            if (!scrypt_set_membuf(&ctx->pass, &ctx->pass_len, p))
+                return 0;
+            break;
+
+        case kdf_scrypt_ctx_param_OSSL_KDF_PARAM_SALT:
+            if (!scrypt_set_membuf(&ctx->salt, &ctx->salt_len, p))
+                return 0;
+            break;
+
+        case kdf_scrypt_ctx_param_OSSL_KDF_PARAM_SCRYPT_N: {
+            uint64_t value;
+
+            if (!OSSL_PARAM_get_uint64(p, &value) {
+                if (!(value > 1 && is_power_of_two(u64_value)))
+                    return 0;
+                ctx->N = value;
+            }
+            break;
+        }
+
+        case kdf_scrypt_ctx_param_OSSL_KDF_PARAM_SCRYPT_R: {
+            uint64_t value;
+
+            if (!OSSL_PARAM_get_uint64(p, &value) {
+                if (!(value >= 1))
+                    return 0;
+                ctx->r = value;
+            }
+            break;
+        }
+
+        case kdf_scrypt_ctx_param_OSSL_KDF_PARAM_SCRYPT_P: {
+            uint64_t value;
+
+            if (!OSSL_PARAM_get_uint64(p, &value) {
+                if (!(value >= 1))
+                    return 0;
+                ctx->p = value;
+            }
+            break;
+        }
+
+        case kdf_scrypt_ctx_param_OSSL_KDF_PARAM_SCRYPT_MAXMEM: {
+            uint64_t value;
+
+            if (!OSSL_PARAM_get_uint64(p, &value) {
+                if (!(value >= 1))
+                    return 0;
+                ctx->p = value;
+            }
+            break;
+        }
+            
+        case kdf_scrypt_ctx_param_OSSL_KDF_PARAM_PROPERTIES:
+            if (p != NULL) {
+                if (p->data_type != OSSL_PARAM_UTF8_STRING) {
+                    if (!set_property_query(ctx, p->data) || !set_digest(ctx))
+                        return 0;
+                    }
+            }
+    }
+
+    return 1;
+}
+
+static const OSSL_PARAM *kdf_scrypt_settable_ctx_params(ossl_unused void *ctx,
+                                                        ossl_unused void *p_ctx)
+{
+    static const OSSL_PARAM known_settable_ctx_params[] = {
+        OSSL_PARAM_octet_string(OSSL_KDF_PARAM_PASSWORD, NULL, 0),
+        OSSL_PARAM_octet_string(OSSL_KDF_PARAM_SALT, NULL, 0),
+        OSSL_PARAM_uint64(OSSL_KDF_PARAM_SCRYPT_N, NULL),
+        OSSL_PARAM_uint32(OSSL_KDF_PARAM_SCRYPT_R, NULL),
+        OSSL_PARAM_uint32(OSSL_KDF_PARAM_SCRYPT_P, NULL),
+        OSSL_PARAM_uint64(OSSL_KDF_PARAM_SCRYPT_MAXMEM, NULL),
+        OSSL_PARAM_utf8_string(OSSL_KDF_PARAM_PROPERTIES, NULL, 0),
+        OSSL_PARAM_END
+    };
+    return known_settable_ctx_params;
+}
+
+static int kdf_scrypt_get_ctx_params(void *vctx, OSSL_PARAM params[])
+{
+    const OSSL_PARAM *p;
+    KDF_SCRYPT *ctx = vctx;
+
+    if (params == NULL)
+        return 1;
+
+    for (p = params; p->key != NULL; p++)
+        switch (kdf_scrypt_ctx_lookup(p))
```

Comment:
nit: {

Reply:
and here.

Still prototype boilerplate code is meant to have errors